homeassistant/components/androidtv/media_player.py

```diff
@@ -99,6 +103,9 @@
         vol.Optional(CONF_APPS, default=dict()): vol.Schema({cv.string: cv.string}),
         vol.Optional(CONF_TURN_ON_COMMAND): cv.string,
         vol.Optional(CONF_TURN_OFF_COMMAND): cv.string,
+        vol.Optional(CONF_STATE_DETECTION_RULES, default=dict()): vol.Schema(
+            {cv.string: ha_state_detection_rules_validator(vol.Invalid)}
```

Comment:
The validator would be more general purpose if it accepted a function instead of an exception. But it's ok like this too.

Reply:
I wasn’t 100% clear on what you meant by your previous post, but I thought this would fit the bill. It's basically the same function as before. I tested it out and it worked as expected, reporting error messages in the "Check config" box when something wasn't right. 

I'd like to leave it as is, since I already tested it and published it on Pypi. But I'm also curious what you meant by your previous post. 

> Even better would be if we moved the whole validator to the library. We can do that by wrapping it in a factory function, that accepts a callback to be called when the validation fails. The callback should accept the message for the failed validation.
>
> Then the user of the library can create a validator by calling the factory passing the callback. In our case the callback will raise `vol.Invalid`.

Reply:
If we have the validator factory accept a function as callback. We can do this:
```py
def handle_invalid_rule(msg):
    raise vol.Invalid(msg)

...

vol.Optional(CONF_STATE_DETECTION_RULES, default=d{}): vol.Schema(
            {cv.string: ha_state_detection_rules_validator(handle_invalid_rule)}
...
```
A function is more flexible since another user of the library might want to do something else than raise an error.

We can keep it like it is.